### DISCUSSION TITLE
[BP#56] Utiliser un système de templating (Smarty, Twig)

### DISCUSSION THREAD

**Discussion by devcrafting** [cnumr/best-practices#207] [Bonnes pratiques] [answered]
A rendre non spécifique PHP et notamment enlever Smarty et Twig du titre

**Comment by devcrafting** [comment]
Je ne connais pas bien l'effet des moteurs de templating dans d'autres languages, quelqu'un a plus d'info?

**Comment by ACTLEM** [comment]
~L~Ces systèmes de templates sont surtout pour PHP, après je ne suis pas spécialiste de PHP donc je ne sais pas si cette BP est encore d'actualité. @arnolem, @yannkozon ou autres personnes spécialistes PHP un avis?

**Reply by tbroyer** [reply]
> Les systèmes de templates sont surtout pour PHP

:interrobang: 

J'utilise Handlebars en Java pour du rendu serveur, j'ai utilisé Closure Templates aussi (on pourrait parler de JSP mais c'est quand même un poil différent) ; Spring Web utilise volontiers FreeMarker ou Thymeleaf.
EcoMeter utilise les templates Go, qu'on s'assure de ne les parser qu'une seule fois: https://gitlab.com/ecoconceptionweb/ecometer/-/blob/901e669804337775e4b6c64d1f12c3c93cea5c0c/main.go#L187-191
Et globalement tous les frameworks du type Ruby on Rails (donc Django en Python, Symfony et autres en PHP, ASP.NET MVC / Razor) utilisent des templates.

J'ai du mal à interpréter la BP par contre: est-ce qu'elle nous dit d'utiliser des systèmes de templating parce que les templates peuvent être précompilés ? (vs. quoi du coup qui ne serait pas précompilé de la même façon ?) est-ce qu'elle nous dit de choisir un système de templating qui supporte ce genre de précompilation et s'assurer qu'il est activé ? (de quelle "précompilation" on parle du coup ? parser/analyser le template une seule fois, vers une structure de données optimisée, et le réutiliser ? ou réellement générer du "code exécutable" ?)

**Reply by ACTLEM** [reply]
:-D...désolé pour ma typo (~L~Ces) et le mot "surtout" qui portent assurément à confusion. On pourrait aussi ajouter Twing, Pug (ex Jade), Haml pour NodeJS. Ma question était surtout d'avoir l'avis de spécialistes PHP sur l'intérêt majeur ou non et si avec les dernières versions de PHP c'est toujours d'actualité sur la BP proprement dite.

Etant donné que je ne connais pas de systèmes de templates qui n'utilisent pas de cache (et je ne pense pas que cela soit le cas sur les systèmes en PHP), et vue l'insistance du cache sur cette BP, je dirais que la BP pousserait à utiliser des systèmes de templates de manière globale.

**Reply by arnolem** [reply]
Déjà, l'exemple basé sur Smarty est obsolète pour moi. Les développeurs PHP vont plutôt utiliser Twig (Symfony) ou Blade (Laravel).
Ensuite, OUI, il faut utiliser la mise en cache lorsqu'on utilise des systèmes de Template. Par exemple, il y a beaucoup de Prestashop qui trainaient des problèmes de performance car le cache Smarty n'était pas activé.

La question serait plutôt de savoir si l'utilisation de moteur de template est bénéfique par rapport à ne pas en utiliser du tout.
Dans des frameworks comme Laminas (Ex Zend Framework), il n'y a pas de moteur de template de base et PHP est lui même utilisé pour la génération des vues. Il n'y a pas besoin d'économiser un temps d'analyse du template puisqu'il n'y en a pas.

Par contre, la BP parle bien d'un cache bytecode (Il voulait peut-être parler d'opcode ?). Ne connaissant pas bien le fonctionnement du cache des systèmes de template, je ne vais pas m'avancer. Soit leur cache leur permet simplement de gommer leurs défauts de lenteur additionnelle, soit leur cache leur fait gagner du temps vis à vis de PHP lui même (j'ai quand même des doutes).

En l'état, je n’affirmerais pas que "Utiliser un système de templating" fait gagner des cycles CPU et s'inscrit dans une démarche d'éco-conception (sans faire des tests sur le sujet ou profiler un programme). Par contre, activer le cache si vous utilisez un système de templating est complétement valide.

**Reply by arnolem** [reply]
Je peux déjà donner un élément de réponse avec un Twig que j'ai sous la main.
Le cache de Twig va simplement créer des fichiers PHP pour chaque fichier de template.
A priori : 
 - il ne va pas limiter les accès disques en réduisant le nombre de fichier à ouvrir depuis le disque
 - il ne va pas faire un cache par "Page de sortie" et éviter des requêtes SQL
 - il ne va pas être meilleur qu'un templating en PHP natif.
 
Donc pour moi, Twig ne fait que compenser son propre temps de traitement grâce à son cache.

**Reply by arnolem** [reply]
Je viens de vérifier avec Laravel et Blade, ça génère également des fichiers PHP qui contiennent du HTML et des `<?php echo $maVariable ?>`
Donc même conclusion pour Blade qui ne ferait pas mieux qu'un template directement en PHP (au niveau de l'éco-conception).

**Reply by arnolem** [reply]
Pour finir, je viens de vérifier sur le Smarty d'un Prestashop et c'est la même chose. Le cache n'est pas en **bytecode** comme indiqué mais simplement en PHP. Donc je pense que cette BP est complétement fausse.

**Reply by arnolem** [reply]
Je suis tombé sur cet article que je n'ai pas lu 😅 mais qui montre la différence de temps de traitement suivant les différents templates.
Aucun moteur de template ne fait mieux que PHP lui même.

![](https://miro.medium.com/max/700/1*VzlJqAd9IUBXf6R_q1Uk0Q.jpeg)

Source : https://medium.com/@gotzmann/the-fastest-template-engine-for-php-b5b461b46296

**Reply by tbroyer** [reply]
…mais un bon moteur de template sera "secure by default" (autoescaping) quand avec "PHP brut" il faudra penser à appeler `htmlspecialchars` à tout bout de champ :wink: 
(j'ai vécu la même chose avec un projet qui utilise JSP, il y a eu des problèmes à cause de ça –heureusement sans conséquence, des apostrophes dans des valeurs qui _fermaient_ un attribut HTML, le reste était donc considéré comme d'autres attributs et donc ignorés puisqu'invalides–, et il y en a encore sûrement qui n'ont juste jamais été détectés)

**Reply by arnolem** [reply]
Oui, c'est pour ça que j'ai précisé que le moteur de template ne fera pas mieux **au niveau de l'éco-conception**. Pour éviter les failles XSS c'est bien entendu le moteur de template que remporte la victoire 😉
Mais comme le livre parle de pratique d'éco-conception, difficile de laisser passer un BP fausse en terme d'éco-conception sous prétexte que ça reste une bonne pratique au sens large.

**Reply by ACTLEM** [reply]
Merci beaucoup @arnolem. Vus les éléments que tu donnes, le templating via Twig ou Smarty n'est finalement pas une bonne pratique d'écoconception en soit...activer le cache doit l'être mais c'est déjà repris par beaucoup de bonnes pratiques existantes.

**Reply by Fred-artwai** [reply]
Effectivement si l'objectif est d'avoir du cache, un système de templating n'est qu'un moyen d'y parvenir. Je ne vois pas l'intérêt de cette BP.

**Reply by tbroyer** [reply]
Note: on parlait ici du "template cache", pas du cache des résultats de l'application du template (qui est effectivement couvert par ailleurs)

**Comment by ACTLEM** [comment, marked as answer]
Vus les avis, on part sur une suppression de la BP, j'ai créé l'issue: https://github.com/cnumr/best-practices/issues/263
